File: tests/python/test_ppo.py
"""
Python binding tests for PPO indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestPpo:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_ppo_partial_params(self, test_data):
        """Test PPO with partial parameters (None values) - mirrors check_ppo_partial_params"""
        close = test_data['close']
        
        # Test with all default params (None)
        result = ta_indicators.ppo(close)  # Using defaults
        assert len(result) == len(close)
    
    def test_ppo_accuracy(self, test_data):
        """Test PPO matches expected values from Rust tests - mirrors check_ppo_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['ppo']
        
        result = ta_indicators.ppo(
            close,
            fast_period=expected['default_params']['fast_period'],
            slow_period=expected['default_params']['slow_period'],
            ma_type=expected['default_params']['ma_type']
        )
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-8,
            msg="PPO last 5 values mismatch"
        )
        
        # Compare full output with Rust
        compare_with_rust('ppo', result, 'close', expected['default_params'])
    
    def test_ppo_default_candles(self, test_data):
        """Test PPO with default parameters - mirrors check_ppo_default_candles"""
        close = test_data['close']
        
        # Default params: fast_period=12, slow_period=26, ma_type='sma'
        result = ta_indicators.ppo(close, fast_period=12, slow_period=26, ma_type='sma')
        assert len(result) == len(close)
    
    def test_ppo_zero_period(self):
        """Test PPO fails with zero period - mirrors check_ppo_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ppo(input_data, fast_period=0, slow_period=26, ma_type='sma')
    
    def test_ppo_period_exceeds_length(self):
        """Test PPO fails when period exceeds data length - mirrors check_ppo_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ppo(data_small, fast_period=12, slow_period=26, ma_type='sma')
    
    def test_ppo_very_small_dataset(self):
        """Test PPO fails with insufficient data - mirrors check_ppo_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ppo(single_point, fast_period=12, slow_period=26, ma_type='sma')
    
    def test_ppo_empty_input(self):
        """Test PPO fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.ppo(empty, fast_period=12, slow_period=26, ma_type='sma')
    
    def test_ppo_nan_handling(self, test_data):
        """Test PPO handles NaN values correctly - mirrors check_ppo_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.ppo(close, fast_period=12, slow_period=26, ma_type='sma')
        assert len(result) == len(close)
        
        # After warmup period (30), no NaN values should exist
        if len(result) > 30:
            assert not np.any(np.isnan(result[30:])), "Found unexpected NaN after warmup period"
    
    def test_ppo_streaming(self, test_data):
        """Test PPO streaming matches batch calculation - mirrors check_ppo_streaming"""
        close = test_data['close']
        fast_period = 12
        slow_period = 26
        ma_type = 'sma'
        
        # Batch calculation
        batch_result = ta_indicators.ppo(close, fast_period=fast_period, slow_period=slow_period, ma_type=ma_type)
        
        # Streaming calculation
        stream = ta_indicators.PpoStream(fast_period=fast_period, slow_period=slow_period, ma_type=ma_type)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        # Compare batch vs streaming
        assert len(batch_result) == len(stream_values)
        
        # Compare values where both are not NaN
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"PPO streaming mismatch at index {i}")
    
    def test_ppo_batch(self, test_data):
        """Test PPO batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.ppo_batch(
            close,
            fast_period_range=(12, 12, 0),  # Default fast period only
            slow_period_range=(26, 26, 0),  # Default slow period only
            ma_type='sma'  # Default ma_type
        )
        
        assert 'values' in result
        assert 'fast_periods' in result
        assert 'slow_periods' in result
        assert 'ma_types' in result
        
        # Check dimensions
        assert result['values'].shape[0] == 1  # Single parameter combination
        assert result['values'].shape[1] == len(close)
        
        # Check parameters
        assert result['fast_periods'][0] == 12
        assert result['slow_periods'][0] == 26
        assert result['ma_types'][0] == 'sma'
        
        # Compare with single calculation
        single_result = ta_indicators.ppo(close, fast_period=12, slow_period=26, ma_type='sma')
        assert_close(
            result['values'][0], 
            single_result, 
            rtol=1e-9,
            msg="PPO batch vs single calculation mismatch"
        )
    
    def test_ppo_batch_multiple_params(self, test_data):
        """Test PPO batch processing with multiple parameter combinations"""
        close = test_data['close']
        
        result = ta_indicators.ppo_batch(
            close,
            fast_period_range=(10, 14, 2),  # 10, 12, 14
            slow_period_range=(24, 28, 2),  # 24, 26, 28
            ma_type='ema'
        )
        
        # Should have 3 * 3 = 9 combinations
        assert result['values'].shape[0] == 9
        assert result['values'].shape[1] == len(close)
        assert len(result['fast_periods']) == 9
        assert len(result['slow_periods']) == 9
        assert len(result['ma_types']) == 9
        
        # Verify parameter combinations
        expected_combinations = [
            (10, 24), (10, 26), (10, 28),
            (12, 24), (12, 26), (12, 28),
            (14, 24), (14, 26), (14, 28)
        ]
        
        for i, (fast, slow) in enumerate(expected_combinations):
            assert result['fast_periods'][i] == fast
            assert result['slow_periods'][i] == slow
            assert result['ma_types'][i] == 'ema'
    
    def test_ppo_kernel_parameter(self, test_data):
        """Test PPO with different kernel parameters"""
        close = test_data['close']
        
        # Test with auto kernel (default)
        result_auto = ta_indicators.ppo(close, kernel=None)
        assert len(result_auto) == len(close)
        
        # Test with scalar kernel
        result_scalar = ta_indicators.ppo(close, kernel='scalar')
        assert len(result_scalar) == len(close)
        
        # Results should be very close regardless of kernel
        assert_close(
            result_auto, 
            result_scalar, 
            rtol=1e-9,
            msg="PPO kernel results mismatch"
        )
    
    def test_ppo_different_ma_types(self, test_data):
        """Test PPO with different moving average types"""
        close = test_data['close']
        
        ma_types = ['sma', 'ema', 'wma']
        results = {}
        
        for ma_type in ma_types:
            results[ma_type] = ta_indicators.ppo(
                close, 
                fast_period=12, 
                slow_period=26, 
                ma_type=ma_type
            )
            assert len(results[ma_type]) == len(close)
        
        # Results should be different for different MA types
        assert not np.array_equal(results['sma'], results['ema'])
        assert not np.array_equal(results['sma'], results['wma'])
        assert not np.array_equal(results['ema'], results['wma'])


if __name__ == "__main__":
    # Run tests in this file
    pytest.main([__file__, "-v"])

File: tests/wasm/test_ppo.js
/**
 * WASM binding tests for PPO indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('PPO partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ppo_js(close, 12, 26, 'sma');
    assert.strictEqual(result.length, close.length);
});

test('PPO accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.ppo;
    
    const result = wasm.ppo_js(
        close,
        expected.defaultParams.fast_period,
        expected.defaultParams.slow_period,
        expected.defaultParams.ma_type
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "PPO last 5 values mismatch"
    );
    
    
    await compareWithRust('ppo', result, 'close', expected.defaultParams);
});

test('PPO default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ppo_js(close, 12, 26, 'sma');
    assert.strictEqual(result.length, close.length);
});

test('PPO zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ppo_js(inputData, 0, 26, 'sma');
    }, /invalid period/i);
});

test('PPO period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ppo_js(dataSmall, 12, 26, 'sma');
    }, /invalid period/i);
});

test('PPO very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.ppo_js(singlePoint, 12, 26, 'sma');
    }, /invalid period/i);
});

test('PPO empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ppo_js(empty, 12, 26, 'sma');
    }, /empty|all values are nan|invalid period/i);
});

test('PPO NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ppo_js(close, 12, 26, 'sma');
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 30) {
        for (let i = 30; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('PPO different MA types', () => {
    
    const close = new Float64Array(testData.close);
    
    const maTypes = ['sma', 'ema', 'wma'];
    const results = {};
    
    for (const maType of maTypes) {
        results[maType] = wasm.ppo_js(close, 12, 26, maType);
        assert.strictEqual(results[maType].length, close.length);
    }
    
    
    assert(!results.sma.every((v, i) => v === results.ema[i]));
    assert(!results.sma.every((v, i) => v === results.wma[i]));
    assert(!results.ema.every((v, i) => v === results.wma[i]));
});

test('PPO batch processing', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        fast_period_range: [12, 12, 0], 
        slow_period_range: [26, 26, 0], 
        ma_type: 'sma'
    };
    
    const result = wasm.ppo_batch(close, config);
    
    assert(result.values);
    assert(result.combos);
    assert.strictEqual(result.rows, 1); 
    assert.strictEqual(result.cols, close.length);
    
    
    assert.strictEqual(result.combos[0].fast_period, 12);
    assert.strictEqual(result.combos[0].slow_period, 26);
    assert.strictEqual(result.combos[0].ma_type, 'sma');
    
    
    const singleResult = wasm.ppo_js(close, 12, 26, 'sma');
    const batchValues = result.values.slice(0, close.length);
    assertArrayClose(
        batchValues,
        singleResult,
        1e-9,
        "PPO batch vs single calculation mismatch"
    );
});

test('PPO batch multiple params', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        fast_period_range: [10, 14, 2], 
        slow_period_range: [24, 28, 2], 
        ma_type: 'ema'
    };
    
    const result = wasm.ppo_batch(close, config);
    
    
    assert.strictEqual(result.rows, 9);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 9);
    
    
    const expectedCombinations = [
        { fast: 10, slow: 24 }, { fast: 10, slow: 26 }, { fast: 10, slow: 28 },
        { fast: 12, slow: 24 }, { fast: 12, slow: 26 }, { fast: 12, slow: 28 },
        { fast: 14, slow: 24 }, { fast: 14, slow: 26 }, { fast: 14, slow: 28 }
    ];
    
    for (let i = 0; i < expectedCombinations.length; i++) {
        assert.strictEqual(result.combos[i].fast_period, expectedCombinations[i].fast);
        assert.strictEqual(result.combos[i].slow_period, expectedCombinations[i].slow);
        assert.strictEqual(result.combos[i].ma_type, 'ema');
    }
});

test('PPO fast API - basic', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const inPtr = wasm.ppo_alloc(len);
    const outPtr = wasm.ppo_alloc(len);
    
    try {
        
        const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
        wasmMemory.set(close, inPtr / 8);
        
        
        wasm.ppo_into(inPtr, outPtr, len, 12, 26, 'sma');
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        const resultCopy = new Float64Array(result); 
        
        
        const safeResult = wasm.ppo_js(close, 12, 26, 'sma');
        assertArrayClose(resultCopy, safeResult, 1e-9, "Fast API result mismatch");
        
    } finally {
        
        wasm.ppo_free(inPtr, len);
        wasm.ppo_free(outPtr, len);
    }
});

test('PPO fast API - in-place (aliasing)', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const ptr = wasm.ppo_alloc(len);
    
    try {
        
        const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
        wasmMemory.set(close, ptr / 8);
        
        
        wasm.ppo_into(ptr, ptr, len, 12, 26, 'sma');
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        const resultCopy = new Float64Array(result); 
        
        
        const safeResult = wasm.ppo_js(close, 12, 26, 'sma');
        assertArrayClose(resultCopy, safeResult, 1e-9, "In-place result mismatch");
        
    } finally {
        
        wasm.ppo_free(ptr, len);
    }
});


if (process.argv[1] === fileURLToPath(import.meta.url)) {
    test.run();
}
